Improve collector documentation

uhoh-itsmaciek opened the pull request:
This is not ready to merge, but wanted to put it out there for review. I documented the immediately obvious settings, and arranged things according to how they are in `config/config.go`. This means that the general settings / additional settings need some re-arranging (e.g., `ignore_schema_regexp` is probably more useful to highlight than `api_base_url`), but I feel like having both sections makes sense since some settings are way less common.

I added question marks where I was not sure about defaults or the description (sometimes there is only a question mark). Some of these I can probably figure out on my own; but I wanted to get the full structure out there first.

I'll add a pganalyze-web PR where we can have a review app.

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+---
+
+The pganalyze collector periodically queries your database and sends metrics and metadata (as "snapshots") to
+the pganalyze app. This is the only component that talks directly to your database. For details on setting up
```
"This is the only component that talks directly to your database" could be emphasized more - maybe we can create a dedicated security section?

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+The pganalyze collector periodically queries your database and sends metrics and metadata (as "snapshots") to
+the pganalyze app. This is the only component that talks directly to your database. For details on setting up
+the collector, see the [collector installation guide](/docs/install/02_installing_the_collector). The collector
+is open-source and the code is available [on GitHub](github.com/pganalyze/collector).
```
this link is missing a leading https://

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+
+# Configuration Settings
+
+The collector can be configured either through a `INI` config file (in a normal install) or environment variables
```
`a` => `an` INI file

lfittl replied:
"normal install" => "package-based install"

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+The collector can be configured either through a `INI` config file (in a normal install) or environment variables
+(typically for running via its Docker image). Most settings can be configured through either mechanism (if both
+are present, the config file takes precedence). Note that a single collector instance can monitor more than one
+database server (though this is not supported when configuring through environment variables). The `[pganalyze]`
```
(though this is not supported when configuring through environment variables)

=> `This is not supported when configuring through environment variables.`

lfittl replied:
And make a new paragraph after this line.

uhoh-itsmaciek replied:
Ended up just dropping the parentheses in the first comment, and adding a comma--I think that reads better and makes it clearer what the "this" is.

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+are present, the config file takes precedence). Note that a single collector instance can monitor more than one
+database server (though this is not supported when configuring through environment variables). The `[pganalyze]`
+section describes settings that apply to all servers, and the other sections are named after the database server
+they correspond to, and define settings for just that server. When using environment variables, the setting applies
```
`sections are named after the database server they correspond to` => we should clarify that its a name you chose (doesn't have to correlate with the server name, or hostname - in fact the example we usually give is `server1` and `server2`)

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+to all monitored servers.
+
+After you make changes, the config file can be reloaded while the collector is running with the command
+`pganalyze-collector --reload`. This minimizes monitoring interruptions and simplifies config file updates.
```
"After you make changes, you can run `pganalyze-collector --test --reload` to verify the new configuration and reload the collector if they work correctly. This minimizes monitoring interruptions and simplifies config file updates."

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+    <tr>
+      <th>setting</th>
+      <th>default</th>
+      <th>description</th>
```
This feels better with uppercase leading letters.

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+    <tr>
+      <td>api_base_url (PGA_API_BASEURL)</td>
+      <td>https://api.pganalyze.com</td>
+      <td>URL to contact the pganalyze app. You typically will not need to change this.</td>
```
Base URL for contacting the pganalyze API. You typically do not need to change this, unless you are running [pganalyze Enterprise Edition](/docs/enterprise).

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+      <td>URL to contact the pganalyze app. You typically will not need to change this.</td>
+    </tr>
+    <tr>
+      <td>error_callback</td>
```
Move to additional settings

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+      </td>
+    </tr>
+    <tr>
+      <td>success_callback</td>
```
Move to additional settings

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+    <tr>
+      <td>enable_reports (PGA_ENABLE_REPORTS)</td>
+      <td>false</td>
+      <td></td>
```
We can just skip this (or comment it out)

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+      <td></td>
+    </tr>
+    <tr>
+      <td>disable_logs (PGA_DISABLE_LOGS)</td>
```
move to additional settings

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+      <td>Disable log collection and processing</td>
+    </tr>
+    <tr>
+      <td>disable_activity (PGA_DISABLE_ACTIVITY)</td>
```
move to additional settings

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+      <td>Disable activity snapshots and processing (current vacuums, queries, etc.)</td>
+    </tr>
+    <tr>
+      <td>enable_log_explain (PGA_ENABLE_LOG_EXPLAIN)</td>
```
this is good to keep in the beginning (as it is now)

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+
+## AWS settings
+
+Only relevant if you are running your database in AWS RDS. Note that the <code>aws_endpoint_*</code>
```
If you use the built-in markdown tags this gets escaped correctly (backticks), otherwise you need to prefix underscores like this: `\_`

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+    <tr>
+      <td>db_name (DB_NAME)</td>
+      <td>n/a - either this or db_url is required</td>
+      <td>name of database to monitor; or, comma-separated list of all databases to monitor,
```
I think it would be good to uppercase the first letter of the description

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+      <td>?</td>
+    </tr>
+    <tr>
+      <td>azure_eventhub_namespace (AZURE_EVENTHUB_NAMESPACE)</td>
```
We could reference the Log Insights documentation here (which explains how to use this setting)

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+  <tbody>
+    <tr>
+      <td>azure_db_server_name (AZURE_DB_SERVER_NAME)</td>
+      <td>n/a, required if azure?</td>
```
Auto-detected from hostname

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+    </tr>
+    <tr>
+      <td>aws_db_instance_id (AWS_INSTANCE_ID)</td>
+      <td>n/a, required if aws</td>
```
Auto-detected from hostname

lfittl replied:
May need to be set manually when using IP addresses, or custom DNS records.

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+  <tbody>
+    <tr>
+      <td>aws_region (AWS_REGION)</td>
+      <td>n/a, required if aws</td>
```
Auto-detected from hostname

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+    </tr>
+    <tr>
+      <td>aws_access_key_id (AWS_ACCESS_KEY_ID)</td>
+      <td>n/a, required if aws and using logs?</td>
```
This is not needed when using instance roles (which we can get from the metadata service when using an EC2 instance)

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+  <tbody>
+    <tr>
+      <td>gcp_cloudsql_instance_id (GCP_CLOUDSQL_INSTANCE_ID)</td>
+      <td>n/a, required if gcp?</td>
```
This may be auto-detected from hostname as well (not sure), double check the code.

uhoh-itsmaciek replied:
Does not look like it, but then I don't see where this happens for AWS either. I searched for [AwsDbInstanceID](https://github.com/pganalyze/collector/search?q=AwsDbInstanceID&unscoped_q=AwsDbInstanceID) and none of those look relevant. :confused:

lfittl replied:
AWS and Azure happen here: 

https://github.com/pganalyze/collector/blob/8bd5b1b56df9423b10bdf284d636fe6ac5e153e7/config/read.go#L166

GCP uses IP addresses instead of hostnames, so the logic doesn't apply there - thus this is required when using Log Insights (and in general recommended, since uniquely identifies the server)

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+    </tr>
+    <tr>
+      <td>gcp_pubsub_subscription (GCP_PUBSUB_SUBSCRIPTION)</td>
+      <td>n/a, required if gcp?</td>
```
Reference log insights docs here

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+    <tr>
+      <td>db_log_location (LOG_LOCATION)</td>
+      <td>?</td>
+      <td>database log file or directory location (must be readable by pganalyze system user)</td>
```
This here could be moved to a dedicated section for self-managed servers.

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+    <tr>
+      <td>db_log_docker_tail</td>
+      <td>[none]</td>
+      <td>experimental: name of docker container to collect logs from using <code>docker logs -t</code></td>
```
This can be documented out for now (we just use it for development purposes)

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+      <td>[none]</td>
+      <td>do not monitor matching tables, schemas, or functions; match is checked against schema-qualified object names (e.g.
+      to ignore table "foo" only in the public schema, set to <code>^public\.foo$</code>)
+      </td>
```
We could move this up into the general section

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,425 @@
+    <tr>
+      <td>filter_query_text (FILTER_QUERY_TEXT)</td>
+      <td>unparsable</td>
+      <td>none/unparsable</td>
```
These settings would be best in a dedicated "PII Filtering" section

uhoh-itsmaciek commented:
@lfittl this probably still needs a few tweaks, and I had some questions in the Additional Settings section, but it could use another look.

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+    <tr>
+      <td>http_proxy (<code>HTTP_PROXY</code>)</td>
+      <td>[none]</td>
+      <td>?</td>
```
Proxy to be used for all HTTP connections, such as API calls. Use for proxies that do not support SSL. Example: `http://username:password@myproxy`

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+    <tr>
+      <td>https_proxy (<code>HTTPS_PROXY</code>)</td>
+      <td>[none]</td>
+      <td>?</td>
```
Proxy to be used for all HTTP connections, such as API calls. Use for proxies that support SSL. Example: `https://username:password@myproxy`

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+    <tr>
+      <td>no_proxy (<code>NO_PROXY</code>)</td>
+      <td>[none]</td>
+      <td>?</td>
```
Comma-delimited list of hostnames that should be accessed directly, without using a configured proxy. Has no effect unless either `HTTP_PROXY` or `HTTPS_PROXY` is specified.

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+    <tr>
+      <td>disable_logs (<code>PGA_DISABLE_LOGS</code>)</td>
+      <td>false</td>
+      <td>Disable log collection and processing</td>
```
Maybe `Disable Log Insights data collection`

lfittl commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+    <tr>
+      <td>disable_activity (<code>PGA_DISABLE_ACTIVITY</code>)</td>
+      <td>false</td>
+      <td>Disable activity snapshots and processing (current vacuums, queries, etc.)</td>
```
`Disable activity snapshot data collection (VACUUM, connection traces, etc.)`

lfittl commented on index.md:
```diff
@@ -82,6 +82,8 @@ title: 'Documentation'
       * [Unique](/docs/explain/other-nodes/unique)
       * [Window Aggregate](/docs/explain/other-nodes/window-aggregate)
 
+* **[Collector](/docs/collector)**
```
Maybe it would be good if we put the settings into their own page, so the `/docs/collector` page only has the introduction notes + security notes, and then links to the configuration settings page.

In the future we should move the monitoring user instructions from the collector README into its own docs page here, and it would make sense if that we have a "Collector overview" page then.

uhoh-itsmaciek commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+
+The collector connects to your database on the port you specify (default 5432) using the standard Postgres
+protocol, to Amazon S3 also over https (port 443) to upload monitoring data, and to the pganalyze API also
+over https. The collector only makes outbound connections&mdash;it does not listen on any ports.
```
That's great, thanks. I'm rewording the Heroku section slightly because the collector app is acting as the drain in this case (the drain in Heroku terminology is what consumes the logs, not the mechanism to get them there). But I think it's great how explicit we can be here.

uhoh-itsmaciek commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+to all monitored servers.
+
+After you make changes, you can run `pganalyze-collector --test --reload` to verify the new configuration and load
+the new configuration in the collector service if they work correctly. This minimizes monitoring interruptions and
```
Yeah, I was going for systemd/upstart terminology here, though I guess that might be confusing in this context. I feel like "background process" is vague but maybe that's fine; I'll change to that to at least make clear it's something local.

uhoh-itsmaciek commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+      <td>enable_log_explain (<code>PGA_ENABLE_LOG_EXPLAIN</code>)</td>
+      <td>false</td>
+      <td>Enable log-based EXPLAIN. See <a href="/docs/explain/setup/log_explain">setup instructions</a>, but note 
+      we recommend using <a href="/docs/explain/setup/auto_explain">autoexplain</a> instead if possible.
```
Though just `auto_explain` since this is in embedded HTML, right?

uhoh-itsmaciek commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+    <tr>
+      <td>db_password (<code>DB_PASSWORD</code>)</td>
+      <td>n/a, either this or db_url is required</td>
+      <td>Password to connect with</td>
```
I wonder if this will confuse people with "Postgres user" vs. "`postgres` user"? But nah, it's probably fine. I'll go with this.

uhoh-itsmaciek commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+    </tr>
+    <tr>
+      <td>db_password (<code>DB_PASSWORD</code>)</td>
+      <td>n/a, either this or db_url is required</td>
```
Or maybe just `[none]` for consistency with other places? I think if you're using the URL you probably know how to specify the password there.

uhoh-itsmaciek commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+    <tr>
+      <td>aws_endpoint_signing_region (<code>AWS_ENDPOINT_SIGNING_REGION</code>)</td>
+      <td>[none]</td>
+      <td>Region to use for signing requests</td>
```
This is very similar to 

```
<td>Path to the client SSL certificate (not needed for most configurations)</td>
```

above, but I like this wording better. I think we should be consistent; I'll make the change here and update that to match.

uhoh-itsmaciek commented on collector/index.md:
```diff
@@ -0,0 +1,498 @@
+
+## Azure settings
+
+Only relevant if you are running your database in Azure. See our <a href="https://pganalyze.com/docs/log-insights/setup/azure-database#step-2-setup-azure-event-hub">Azure setup instructions</a> for details.
```
Good catch--this was initially inline in the table and I hoisted it for clarity. I'll drop the fragment.

uhoh-itsmaciek commented:
@lfittl thanks for the thorough feedback. I addressed all your comments and split the settings into their own page. I also adjusted the heading sizes to patch the other pages (I noticed we don't use `#` because that's the same size as the frontmatter `title`), and moved the settings title down a bit since most pages seem to have a brief intro to avoid two headings in a row.

lfittl commented on collector/settings.md:
```diff
@@ -0,0 +1,499 @@
+interruptions and simplifies config file updates.
+
+
+## Configuration settings
```
This heading also shows up in the TOC, which looks a bit odd. I wonder if we can just skip the heading, and make it one large intro section?

uhoh-itsmaciek replied:
Yeah, you're right--I think that works better.

lfittl reviewed (approved):
Perfect - found one more thing, but otherwise lets 🚢
